File: patterns/02-reliability-patterns/06-hedged-requests/go/hedged_requests.go
package hedgedrequests

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type candidate struct {
	time int
	name string
}

func Evaluate(input map[string]any) (any, error) {
	if input["idempotent"] != true {
		return nil, &PatternError{code: "hedge_not_allowed"}
	}
	primary := input["primary"].(map[string]any)
	delay := int(input["hedge_delay_ms"].(float64))
	primaryLatency := int(primary["latency_ms"].(float64))
	if primary["result"] == "success" && primaryLatency <= delay {
		return map[string]any{"winner": "primary", "completed_ms": primaryLatency, "attempts": 1, "cancelled": []string{}}, nil
	}
	start := delay
	if primary["result"] == "failure" && primaryLatency < start {
		start = primaryLatency
	}
	hedge := input["hedge"].(map[string]any)
	values := []candidate{}
	if primary["result"] == "success" {
		values = append(values, candidate{primaryLatency, "primary"})
	}
	if hedge["result"] == "success" {
		values = append(values, candidate{start + int(hedge["latency_ms"].(float64)), "hedge"})
	}
	if len(values) == 0 {
		return nil, &PatternError{code: "all_attempts_failed"}
	}
	winner := values[0]
	for _, value := range values[1:] {
		if value.time < winner.time {
			winner = value
		}
	}
	cancelled := []string{}
	for _, value := range values {
		if value.name != winner.name && winner.time < value.time {
			cancelled = append(cancelled, value.name)
		}
	}
	return map[string]any{"winner": winner.name, "completed_ms": winner.time, "attempts": 2, "cancelled": cancelled}, nil
}

File: patterns/02-reliability-patterns/06-hedged-requests/python/hedged_requests.py
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
def evaluate(input_data):
    if not input_data.get("idempotent",False):raise PatternError("hedge_not_allowed")
    primary=input_data["primary"]
    if primary["result"]=="success" and primary["latency_ms"]<=input_data["hedge_delay_ms"]:return{"winner":"primary","completed_ms":primary["latency_ms"],"attempts":1,"cancelled":[]}
    start=min(input_data["hedge_delay_ms"],primary["latency_ms"] if primary["result"]=="failure" else input_data["hedge_delay_ms"]);hedge=input_data["hedge"];candidates=[]
    if primary["result"]=="success":candidates.append((primary["latency_ms"],"primary"))
    if hedge["result"]=="success":candidates.append((start+hedge["latency_ms"],"hedge"))
    if not candidates:raise PatternError("all_attempts_failed")
    completed,winner=min(candidates);cancelled=[name for _,name in candidates if name!=winner and completed<dict((name,time) for time,name in candidates)[name]];return{"winner":winner,"completed_ms":completed,"attempts":2,"cancelled":cancelled}
